
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Automatic documentation from sources, for MkDocs.">
      
      
      
        <link rel="canonical" href="https://gnaprs.github.io/scpviz/tutorials/quickstart/">
      
      
        <link rel="prev" href="../installation/">
      
      
        <link rel="next" href="../single_cell/">
      
      
        
      
      
      <link rel="icon" href="../../assets/300ppi/logo_no_text@300x.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Quickstart - scpviz</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#quickstart" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="scpviz" class="md-header__button md-logo" aria-label="scpviz" data-md-component="logo">
      
  <img src="../../assets/300ppi/logo_no_text@300x.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            scpviz
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Quickstart
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="deep-orange"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/gnaprs/scpviz" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    gnaprs/scpviz
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../installation/" class="md-tabs__link">
          
  
  
    
  
  Getting started

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../reference/" class="md-tabs__link">
          
  
  
    
  
  API reference

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../dev/contributing/" class="md-tabs__link">
          
  
  
    
  
  Development

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../admonitions/" class="md-tabs__link">
          
  
  
    
  
  Others

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="https://gnaprs.github.io/" class="md-tabs__link">
        
  
  
    
  
  Author's website

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="scpviz" class="md-nav__button md-logo" aria-label="scpviz" data-md-component="logo">
      
  <img src="../../assets/300ppi/logo_no_text@300x.png" alt="logo">

    </a>
    scpviz
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/gnaprs/scpviz" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    gnaprs/scpviz
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../.." class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_1" id="__nav_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../CHANGELOG/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Changelog
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Getting started
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Getting started
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Quickstart
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Quickstart
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#import" class="md-nav__link">
    <span class="md-ellipsis">
      
        Import
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-processing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Pre-processing
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Pre-processing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filtering" class="md-nav__link">
    <span class="md-ellipsis">
      
        Filtering
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quick visualization
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#normalization-and-imputation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Normalization and Imputation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#differential-expression" class="md-nav__link">
    <span class="md-ellipsis">
      
        Differential expression
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Differential expression">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#volcano-plots" class="md-nav__link">
    <span class="md-ellipsis">
      
        Volcano Plots
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#string-enrichment" class="md-nav__link">
    <span class="md-ellipsis">
      
        STRING enrichment
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      
        Next steps
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../single_cell/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Single-cell Example
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_4" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    
  
    In-depth Tutorials
  

    
  </span>
  
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_2_4" id="__nav_2_4_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    In-depth Tutorials
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../importing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1. Importing Data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2. Filtering
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../imputation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    3. Imputation + Normalization
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    4. Plotting
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../de/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    5. Differential Expression
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../enrichment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    6. Enrichment + Networks
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../reference/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    
  
    API reference
  

    
  </span>
  
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    API reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    pAnnData
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    pAnnData
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/pAnnData/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Base Class
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/import_mixin/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Importing
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/core_mixins/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Core Utilities
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/metadata_mixins/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Metadata + Summary
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/editing_mixins/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Editing + Filtering
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/analysis_mixins/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Analysis + Enrichment
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/identifier_mixins/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Identifiers + STRING
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/plot_mixin/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Visualization
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" >
        
          
          <label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Modules
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Modules
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    scpviz.utils
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    scpviz.plotting
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" >
        
          
          <label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Developer Utilities
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Developer Utilities
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/TrackedDataFrame/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TrackedDataFrame
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reference/pAnnData/hidden_functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Hidden Functions
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Development
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Development
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../dev/contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Contributing
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../dev/CODE_OF_CONDUCT/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code of Conduct
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Others
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Others
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../admonitions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Admonitions
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../content-tabs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Content-Tabs
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../diagram/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Diagram
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="https://gnaprs.github.io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Author's website
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#import" class="md-nav__link">
    <span class="md-ellipsis">
      
        Import
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-processing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Pre-processing
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Pre-processing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filtering" class="md-nav__link">
    <span class="md-ellipsis">
      
        Filtering
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quick visualization
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#normalization-and-imputation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Normalization and Imputation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#differential-expression" class="md-nav__link">
    <span class="md-ellipsis">
      
        Differential expression
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Differential expression">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#volcano-plots" class="md-nav__link">
    <span class="md-ellipsis">
      
        Volcano Plots
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#string-enrichment" class="md-nav__link">
    <span class="md-ellipsis">
      
        STRING enrichment
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      
        Next steps
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
                




              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="quickstart">Quickstart</h1>
<p><a href="https://github.com/gnaprs/scpviz/raw/main/docs/tutorials/quickstart.ipynb"><img alt="Download Notebook" src="https://img.shields.io/badge/download-quickstart__notebook-blue?logo=icloud&amp;style=flat-square" /></a>
<a href="https://colab.research.google.com/github/gnaprs/scpviz/blob/main/docs/tutorials/quickstart.ipynb"><img alt="Open In Colab" src="https://img.shields.io/badge/open%20in-colab-yellow?logo=googlecolab&amp;style=flat-square" /></a></p>
<p>This quickstart tutorial demonstrates a minimal end-to-end workflow using <strong>scpviz</strong>.</p>
<p>First, install <code>scpviz</code> and import the modules:</p>
<div class="language-py highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="n">scpviz</span>
</span></code></pre></div>
<div class="language-py highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="kn">from</span><span class="w"> </span><span class="nn">scpviz</span><span class="w"> </span><span class="kn">import</span> <span class="n">pAnnData</span> <span class="k">as</span> <span class="n">pAnnData</span>
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="kn">from</span><span class="w"> </span><span class="nn">scpviz</span><span class="w"> </span><span class="kn">import</span> <span class="n">plotting</span> <span class="k">as</span> <span class="n">scplt</span>
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="kn">from</span><span class="w"> </span><span class="nn">scpviz</span><span class="w"> </span><span class="kn">import</span> <span class="n">utils</span> <span class="k">as</span> <span class="n">scutils</span>
</span></code></pre></div>
<hr />
<h2 id="import">Import</h2>
<p>Proteomics data is stored in a <code>pAnnData</code> object. <code>scpviz</code> currently supports two data formats: <strong>Proteome Discoverer</strong> (Thermo Fisher) and <strong>DIA-NN</strong> reports. Weâ€™ve provided sample test files below so you can follow along:</p>
<p><a href="https://github.com/gnaprs/scpviz/raw/main/docs/assets/pd32_Proteins.txt"><img alt="Download PD3.2 Proteins File" src="https://img.shields.io/badge/download-pd32__Proteins.txt-blue?logo=icloud&amp;style=flat-square" /></a>
<a href="https://github.com/gnaprs/scpviz/raw/main/docs/assets/pd32_PeptideSequenceGroups.txt"><img alt="Download PD3.2 Peptides File" src="https://img.shields.io/badge/download-pd32__PeptideSequenceGroups.txt-blue?logo=icloud&amp;style=flat-square" /></a>
<a href="https://github.com/gnaprs/scpviz/raw/main/docs/assets/diann_report.parquet"><img alt="Download DIA-NN Report File" src="https://img.shields.io/badge/download-diann__report.parquet-blue?logo=icloud&amp;style=flat-square" /></a></p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Proteome Discoverer</label><label for="__tabbed_1_2">DIA-NN</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-py highlight"><span class="filename">Import Proteome Discoverer data</span><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="n">obs_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span><span class="s1">&#39;cellline&#39;</span><span class="p">,</span><span class="s1">&#39;treatment&#39;</span><span class="p">,</span><span class="s1">&#39;condition&#39;</span><span class="p">,</span><span class="s1">&#39;duration&#39;</span><span class="p">]</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="n">pdata</span> <span class="o">=</span> <span class="n">pAnnData</span><span class="o">.</span><span class="n">import_data</span><span class="p">(</span>
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>    <span class="n">source_type</span><span class="o">=</span><span class="s1">&#39;pd&#39;</span><span class="p">,</span> 
</span><span id="__span-2-4"><a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>    <span class="n">prot_file</span> <span class="o">=</span> <span class="s1">&#39;pd_prot.txt&#39;</span><span class="p">,</span> 
</span><span id="__span-2-5"><a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>    <span class="n">pep_file</span><span class="o">=</span><span class="s1">&#39;pd32_PeptideSequenceGroups.txt&#39;</span><span class="p">,</span>
</span><span id="__span-2-6"><a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>    <span class="n">obs_columns</span><span class="o">=</span><span class="n">obs_columns</span><span class="p">)</span>
</span></code></pre></div>
<div class="result" markdown>
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>ðŸ§­ [USER] Importing data of type [pd]
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>--------------------------
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>Starting import [Proteome Discoverer]
</span><span id="__span-3-4"><a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>
</span><span id="__span-3-5"><a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>Source file: ../assets/pd32_Proteins.txt / ../assets/pd32_PeptideSequenceGroups.txt
</span><span id="__span-3-6"><a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>Number of files: 12
</span><span id="__span-3-7"><a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>Proteins: 10393
</span><span id="__span-3-8"><a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>Peptides: 167114
</span><span id="__span-3-9"><a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a>    ...
</span><span id="__span-3-10"><a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a>    âœ… [OK] pAnnData object is valid.
</span><span id="__span-3-11"><a id="__codelineno-3-11" name="__codelineno-3-11" href="#__codelineno-3-11"></a>    âœ… [OK] Import complete. Use `print(pdata)` to view the object.
</span><span id="__span-3-12"><a id="__codelineno-3-12" name="__codelineno-3-12" href="#__codelineno-3-12"></a>--------------------------
</span></code></pre></div>
</div>
</div>
<div class="tabbed-block">
<div class="language-py highlight"><span class="filename">Import DIA-NN data</span><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="n">obs_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;user&#39;</span><span class="p">,</span> <span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;ms&#39;</span><span class="p">,</span> <span class="s1">&#39;acquisition&#39;</span><span class="p">,</span> <span class="s1">&#39;faims&#39;</span><span class="p">,</span> <span class="s1">&#39;column&#39;</span><span class="p">,</span> <span class="s1">&#39;gradient&#39;</span><span class="p">,</span> <span class="s1">&#39;amount&#39;</span><span class="p">,</span> <span class="s1">&#39;region&#39;</span><span class="p">,</span> <span class="s1">&#39;rep&#39;</span><span class="p">]</span>
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="n">pdata</span> <span class="o">=</span> <span class="n">pAnnData</span><span class="o">.</span><span class="n">import_data</span><span class="p">(</span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>    <span class="n">source_type</span><span class="o">=</span><span class="s1">&#39;diann&#39;</span><span class="p">,</span> 
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>    <span class="n">report_file</span> <span class="o">=</span> <span class="s1">&#39;diann_report.parquet&#39;</span><span class="p">,</span> 
</span><span id="__span-4-5"><a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>    <span class="n">obs_columns</span><span class="o">=</span><span class="n">obs_columns</span><span class="p">)</span>
</span></code></pre></div>
<div class="result" markdown>
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>ðŸ§­ [USER] Importing data of type [diann]
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>--------------------------
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>Starting import [DIA-NN]
</span><span id="__span-5-4"><a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a>
</span><span id="__span-5-5"><a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>Source file: ../assets/diann_report.parquet
</span><span id="__span-5-6"><a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a>Number of files: 12
</span><span id="__span-5-7"><a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a>Proteins: 12652
</span><span id="__span-5-8"><a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a>Peptides: 251047
</span><span id="__span-5-9"><a id="__codelineno-5-9" name="__codelineno-5-9" href="#__codelineno-5-9"></a>    ...
</span><span id="__span-5-10"><a id="__codelineno-5-10" name="__codelineno-5-10" href="#__codelineno-5-10"></a>    âœ… [OK] pAnnData object is valid.
</span><span id="__span-5-11"><a id="__codelineno-5-11" name="__codelineno-5-11" href="#__codelineno-5-11"></a>    âœ… [OK] Import complete. Use `print(pdata)` to view the object.
</span><span id="__span-5-12"><a id="__codelineno-5-12" name="__codelineno-5-12" href="#__codelineno-5-12"></a>--------------------------
</span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Refer to the <a href="../importing/">Importing Data</a> tutorial for more options, including defining metadata columns.</p>
</div>
<p>We can have a quick look at our <code>pdata</code> object by checking <code>pdata.summary</code>:</p>
<div class="language-py highlight"><span class="filename">Checking pdata.summary</span><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="n">pdata</span><span class="o">.</span><span class="n">summary</span>
</span></code></pre></div>
<div class="result">
<table>
<thead>
<tr>
<th style="text-align: center;">Sample</th>
<th style="text-align: center;">cellline</th>
<th style="text-align: center;">treatment</th>
<th style="text-align: center;">condition</th>
<th style="text-align: center;">duration</th>
<th style="text-align: center;">...</th>
<th style="text-align: right;">protein_count</th>
<th style="text-align: center;">...</th>
<th style="text-align: right;">peptide_count</th>
<th style="text-align: center;">...</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">F4</td>
<td style="text-align: center;">AS</td>
<td style="text-align: center;">RA</td>
<td style="text-align: center;">kd</td>
<td style="text-align: center;">d7</td>
<td style="text-align: center;">â€¦</td>
<td style="text-align: right;">9660</td>
<td style="text-align: center;">â€¦</td>
<td style="text-align: right;">139793</td>
<td style="text-align: center;">â€¦</td>
</tr>
<tr>
<td style="text-align: center;">F23</td>
<td style="text-align: center;">BE</td>
<td style="text-align: center;">RA</td>
<td style="text-align: center;">sc</td>
<td style="text-align: center;">d7</td>
<td style="text-align: center;">â€¦</td>
<td style="text-align: right;">1724</td>
<td style="text-align: center;">â€¦</td>
<td style="text-align: right;">4534</td>
<td style="text-align: center;">â€¦</td>
</tr>
<tr>
<td style="text-align: center;">F24</td>
<td style="text-align: center;">BE</td>
<td style="text-align: center;">RA</td>
<td style="text-align: center;">sc</td>
<td style="text-align: center;">d7</td>
<td style="text-align: center;">â€¦</td>
<td style="text-align: right;">9748</td>
<td style="text-align: center;">â€¦</td>
<td style="text-align: right;">133675</td>
<td style="text-align: center;">â€¦</td>
</tr>
</tbody>
</table>
</div>
<hr />
<h2 id="pre-processing">Pre-processing</h2>
<h3 id="filtering">Filtering</h3>
<p>File <strong>F23</strong> shows a markedly low protein count (~1700) compared to ~9000 in all other samples.
This file is an outlier and should be removed prior to downstream analysis.</p>
<p>Two filtering approaches can be used to achieve the same result:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">Filter by protein count</label><label for="__tabbed_2_2">Exclude by filename</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-py highlight"><span class="filename">Filter by minimum protein count</span><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="n">pdata</span> <span class="o">=</span> <span class="n">pdata</span><span class="o">.</span><span class="n">filter_sample</span><span class="p">(</span><span class="n">min_prot</span><span class="o">=</span><span class="mi">8000</span><span class="p">)</span>
</span></code></pre></div>
<div class="result" markdown>
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>ðŸ§­ [USER] Filtering samples [condition]:
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>    Returning a copy of sample data based on condition:
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>    ðŸ”¸ Condition: protein_count &gt;= 8000
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>    â„¹ï¸ Auto-cleanup: No empty proteins found (all-NaN or all-zero).
</span><span id="__span-8-5"><a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>    â†’ Samples kept: 11, Samples dropped: 1
</span><span id="__span-8-6"><a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>    â†’ Proteins kept: 10393
</span></code></pre></div>
</div>
</div>
<div class="tabbed-block">
<div class="language-py highlight"><span class="filename">Filter by excluding specific file</span><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="n">pdata</span> <span class="o">=</span> <span class="n">pdata</span><span class="o">.</span><span class="n">filter_sample</span><span class="p">(</span><span class="n">exclude_file_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;F23&#39;</span><span class="p">])</span>
</span></code></pre></div>
<div class="result" markdown>
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>ðŸ§­ [USER] Filtering samples [file list]:
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>    Returning a copy of sample data based on file list:
</span><span id="__span-10-3"><a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a>    ðŸ”¸ Files requested (excluding): 1
</span><span id="__span-10-4"><a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a>    â„¹ï¸ Auto-cleanup: No empty proteins found (all-NaN or all-zero).
</span><span id="__span-10-5"><a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a>    â†’ Samples kept: 11, Samples dropped: 1
</span><span id="__span-10-6"><a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a>    â†’ Proteins kept: 10393
</span></code></pre></div>
</div>
</div>
</div>
</div>
<p>After filtering, the dataset now contains <strong>11 samples</strong>.  </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All filter functions return a <strong>copy</strong> of the <code>pAnnData</code> object unless <code>inplace=True</code> is specified.<br />
See the <a href="../filtering/">Filtering</a> tutorial for more options.</p>
</div>
<h3 id="quick-visualization">Quick visualization</h3>
<p>Visualize proteins of interest across samples:</p>
<div class="language-py highlight"><span class="filename">Plot protein abundance</span><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>
</span><span id="__span-11-3"><a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</span><span id="__span-11-4"><a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="n">pdata</span><span class="o">.</span><span class="n">plot_abundance</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">namelist</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GAPDH&quot;</span><span class="p">,</span> <span class="s2">&quot;VCP&quot;</span><span class="p">,</span> <span class="s2">&quot;AHNAK&quot;</span><span class="p">],</span> <span class="n">classes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cellline&quot;</span><span class="p">,</span><span class="s2">&quot;condition&quot;</span><span class="p">])</span>
</span><span id="__span-11-5"><a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</span></code></pre></div>
<div class="result">
<figure>
<img alt="Protein abundance plot" src="../../assets/quickstart_1.png" />

<figcaption>Protein abundance visualization for selected proteins.</figcaption>
</figure>
</div>
<hr />
<p>Alternatively, we can explore the PCA embeddings to get an overview of sample clustering.
Other dimensionality reduction methods such as UMAP and t-SNE can also be used.</p>
<div class="language-py highlight"><span class="filename">Plot PCA embeddings</span><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="n">ax</span> <span class="o">=</span> <span class="n">scplt</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">pdata</span><span class="p">,</span> <span class="n">classes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cellline&quot;</span><span class="p">,</span><span class="s2">&quot;condition&quot;</span><span class="p">],</span> <span class="n">add_ellipses</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span></code></pre></div>
<div class="result">
<figure>
  <img alt="Protein abundance plot" src="../../assets/quickstart_2.png" />
<br />
<figcaption></figcaption>
</figure>
</div>
<p>The samples appear well-clustered by both cell line and condition, suggesting good reproducibility and biological separation.</p>
<hr />
<p>Finally, we can examine the coefficient of variation (CV) of each sample.
Here, the samples show overall low variability (median ~0.1), with slightly higher CVs observed in the AS_sc group (~0.35).</p>
<div class="language-py highlight"><span class="filename">Plot sample CVs</span><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</span><span id="__span-13-2"><a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="n">ax</span> <span class="o">=</span> <span class="n">scplt</span><span class="o">.</span><span class="n">plot_cv</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">pdata</span><span class="p">,</span> <span class="n">classes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cellline&quot;</span><span class="p">,</span><span class="s2">&quot;condition&quot;</span><span class="p">])</span>
</span></code></pre></div>
<div class="result">
<figure>
  <img alt="Sample CV plot" src="../../assets/quickstart_3.png" />
<br />
<figcaption></figcaption>
</figure>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Refer to the <a href="../plotting/">plotting</a> tutorial for more advanced plotting options.</p>
</div>
<hr />
<h3 id="normalization-and-imputation">Normalization and Imputation</h3>
<p>Biological and technical variation across samples (e.g., in the AS_sc group) can arise from sample processing or data acquisition. Normalize your data to reduce variation between samples â€” for example, using median scaling.  </p>
<div class="language-py highlight"><span class="filename">Normalization and imputation</span><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="n">pdata</span><span class="o">.</span><span class="n">normalize</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;median&quot;</span><span class="p">)</span>
</span><span id="__span-14-2"><a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="n">pdata</span><span class="o">.</span><span class="n">impute</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;min&quot;</span><span class="p">)</span>
</span></code></pre></div>
<div class="result">
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>ðŸ§­ [USER] Global normalization using &#39;median&#39;. Layer will be saved as &#39;X_norm_median&#39;.
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>     âœ… Normalized all 11 samples.
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>     â„¹ï¸ Set protein data to layer X_norm_median.
</span><span id="__span-15-4"><a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a>ðŸ§­ [USER] Global imputation using &#39;min&#39;. Layer saved as &#39;X_impute_min&#39;. Minimum scaled by 1.
</span><span id="__span-15-5"><a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a>     âœ… 8234 values imputed.
</span><span id="__span-15-6"><a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>     â„¹ï¸ 11 samples fully imputed, 0 samples partially imputed, 0 skipped feature(s) with all missing values.
</span><span id="__span-15-7"><a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a>     â„¹ï¸ Set protein data to layer X_impute_min.
</span></code></pre></div>
</div>
<p>After normalization, CVs for the <strong>AS_sc</strong> group improve compared to pre-normalized data.</p>
<div class="language-py highlight"><span class="filename">Plot sample CVs after normalization</span><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="n">ax</span> <span class="o">=</span> <span class="n">scplt</span><span class="o">.</span><span class="n">plot_cv</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">pdata</span><span class="p">,</span> <span class="n">classes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cellline&quot;</span><span class="p">,</span><span class="s2">&quot;condition&quot;</span><span class="p">])</span>
</span></code></pre></div>
<div class="result">
<figure>
  <img alt="Sample CV plot" src="../../assets/quickstart_4.png" />
<br />
<figcaption>Sample-wise CVs after normalization and imputation.</figcaption>
</figure>
</div>
<p>Other imputation methods are also available, including KNN, median, and minimum with a scaling factor. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Refer to the <a href="../imputation/">Normalization &amp; Imputation</a> tutorial for additional examples and parameter options, such as Harmony or DirectLFQ.</p>
</div>
<h2 id="differential-expression">Differential expression</h2>
<h3 id="volcano-plots">Volcano Plots</h3>
<p>Run a <strong>differential expression (DE)</strong> analysis, commonly visualized with volcano plots. To start, we define a <em>comparison ratio</em>: for instance, comparing cell line BE under the kd condition against cell line BE under sc.</p>
<div class="language-py highlight"><span class="filename">Differentiatial expression with volcano plots</span><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a><span class="n">comparison_values</span><span class="o">=</span><span class="p">[{</span><span class="s1">&#39;cellline&#39;</span><span class="p">:</span><span class="s1">&#39;BE&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">:</span><span class="s1">&#39;kd&#39;</span><span class="p">},{</span><span class="s1">&#39;cellline&#39;</span><span class="p">:</span><span class="s1">&#39;BE&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">:</span><span class="s1">&#39;sc&#39;</span><span class="p">}]</span>
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="n">ax</span> <span class="o">=</span> <span class="n">scplt</span><span class="o">.</span><span class="n">plot_volcano</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">pdata</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">comparison_values</span><span class="p">)</span>
</span></code></pre></div>
<div class="result">
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>ðŸ§­ [USER] Running differential expression [protein]
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a>   ðŸ”¸ Comparing groups: [{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;kd&#39;}] vs [{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;sc&#39;}]
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a>   ðŸ”¸ Group sizes: 3 vs 2 samples
</span><span id="__span-18-4"><a id="__codelineno-18-4" name="__codelineno-18-4" href="#__codelineno-18-4"></a>   ðŸ”¸ Method: ttest | Fold Change: mean | Layer: X
</span><span id="__span-18-5"><a id="__codelineno-18-5" name="__codelineno-18-5" href="#__codelineno-18-5"></a>   ðŸ”¸ P-value threshold: 0.05 | Log2FC threshold: 1
</span><span id="__span-18-6"><a id="__codelineno-18-6" name="__codelineno-18-6" href="#__codelineno-18-6"></a>     âœ… DE complete. Results stored in:
</span><span id="__span-18-7"><a id="__codelineno-18-7" name="__codelineno-18-7" href="#__codelineno-18-7"></a><span class="hll">       â€¢ .stats[&quot;[{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;kd&#39;}] vs [{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;sc&#39;}]&quot;]
</span></span><span id="__span-18-8"><a id="__codelineno-18-8" name="__codelineno-18-8" href="#__codelineno-18-8"></a>       â€¢ Columns: log2fc, p_value, significance, etc.
</span><span id="__span-18-9"><a id="__codelineno-18-9" name="__codelineno-18-9" href="#__codelineno-18-9"></a>       â€¢ Upregulated: 279 | Downregulated: 258 | Not significant: 9856
</span></code></pre></div>
</div>
<div class="result">
<figure>
  <img alt="Volcano plot" src="../../assets/quickstart_5.png" />
<br />
<figcaption>Volcano plot highlighting upregulated and downregulated proteins.</figcaption>
</figure>
</div>
<p>Access the DE results stored in <code>.stats</code> under the key shown in the output.</p>
<div class="language-py highlight"><span class="filename">Access DE results</span><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a><span class="n">pdata</span><span class="o">.</span><span class="n">stats</span><span class="p">[</span><span class="s2">&quot;[{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;kd&#39;}] vs [{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;sc&#39;}]&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
</span></code></pre></div>
<div class="result">
<table>
<thead>
<tr>
<th style="text-align: left;">Genes</th>
<th style="text-align: right;">[{'cellline': 'BE', 'condition': 'kd'}]</th>
<th style="text-align: right;">[{'cellline': 'BE', 'condition': 'sc'}]</th>
<th style="text-align: right;">log2fc</th>
<th style="text-align: right;">p_value</th>
<th style="text-align: right;">test_statistic</th>
<th style="text-align: right;">significance_score</th>
<th style="text-align: left;">significance</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">PPP1R37</td>
<td style="text-align: right;">601891.3155</td>
<td style="text-align: right;">103551.4786</td>
<td style="text-align: right;">2.54</td>
<td style="text-align: right;">0.0118</td>
<td style="text-align: right;">5.51</td>
<td style="text-align: right;">4.90</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">IGSF9B</td>
<td style="text-align: right;">438967.4093</td>
<td style="text-align: right;">193638.7087</td>
<td style="text-align: right;">1.18</td>
<td style="text-align: right;">0.0159</td>
<td style="text-align: right;">4.94</td>
<td style="text-align: right;">2.12</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">GPR161</td>
<td style="text-align: right;">126213.0252</td>
<td style="text-align: right;">54380.4809</td>
<td style="text-align: right;">1.21</td>
<td style="text-align: right;">0.0102</td>
<td style="text-align: right;">5.81</td>
<td style="text-align: right;">2.42</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">TIGD5</td>
<td style="text-align: right;">43568.9795</td>
<td style="text-align: right;">9048.0415</td>
<td style="text-align: right;">2.27</td>
<td style="text-align: right;">0.0230</td>
<td style="text-align: right;">4.31</td>
<td style="text-align: right;">3.71</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">TTC9B</td>
<td style="text-align: right;">222577.1287</td>
<td style="text-align: right;">26482.7005</td>
<td style="text-align: right;">3.07</td>
<td style="text-align: right;">1.81e-05</td>
<td style="text-align: right;">49.58</td>
<td style="text-align: right;">14.57</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">NMNAT2</td>
<td style="text-align: right;">269130.1753</td>
<td style="text-align: right;">82171.6244</td>
<td style="text-align: right;">1.71</td>
<td style="text-align: right;">0.0046</td>
<td style="text-align: right;">7.69</td>
<td style="text-align: right;">4.01</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">ATXN7L1</td>
<td style="text-align: right;">254803.2800</td>
<td style="text-align: right;">66922.2053</td>
<td style="text-align: right;">1.93</td>
<td style="text-align: right;">0.0123</td>
<td style="text-align: right;">5.42</td>
<td style="text-align: right;">3.68</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">SASS6</td>
<td style="text-align: right;">1765918.661</td>
<td style="text-align: right;">779926.980</td>
<td style="text-align: right;">1.18</td>
<td style="text-align: right;">0.0365</td>
<td style="text-align: right;">3.61</td>
<td style="text-align: right;">1.69</td>
<td style="text-align: left;">upregulated</td>
</tr>
<tr>
<td style="text-align: left;">â€¦</td>
<td style="text-align: right;">â€¦</td>
<td style="text-align: right;">â€¦</td>
<td style="text-align: right;">â€¦</td>
<td style="text-align: right;">â€¦</td>
<td style="text-align: right;">â€¦</td>
<td style="text-align: right;">â€¦</td>
<td style="text-align: left;">â€¦</td>
</tr>
</tbody>
</table>
<p>The table above shows the top DE results (<code>df.head(8)</code>) including <strong>logâ‚‚ fold change</strong>, <strong>p-value</strong>, and <strong>significance score</strong>.</p>
</div>
<hr />
<h3 id="string-enrichment">STRING enrichment</h3>
<p>We can perform <strong>STRING enrichment</strong> on the sets of up- and downregulated proteins from our DE analysis. First, list the available enrichment keys:</p>
<div class="language-py highlight"><span class="filename">List enrichment keys</span><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a><span class="n">pdata</span><span class="o">.</span><span class="n">list_enrichments</span><span class="p">()</span>
</span></code></pre></div>
<div class="result">
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>ðŸ§­ [USER] Listing STRING enrichment status
</span><span id="__span-21-2"><a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a>
</span><span id="__span-21-3"><a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a>     â„¹ï¸ Available DE comparisons (not yet enriched):
</span><span id="__span-21-4"><a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a><span class="hll">        - BE_kd vs BE_sc
</span></span><span id="__span-21-5"><a id="__codelineno-21-5" name="__codelineno-21-5" href="#__codelineno-21-5"></a>
</span><span id="__span-21-6"><a id="__codelineno-21-6" name="__codelineno-21-6" href="#__codelineno-21-6"></a>  ðŸ”¹ To run enrichment:
</span><span id="__span-21-7"><a id="__codelineno-21-7" name="__codelineno-21-7" href="#__codelineno-21-7"></a>      pdata.enrichment_functional(from_de=True, de_key=&quot;...&quot;)
</span><span id="__span-21-8"><a id="__codelineno-21-8" name="__codelineno-21-8" href="#__codelineno-21-8"></a>
</span><span id="__span-21-9"><a id="__codelineno-21-9" name="__codelineno-21-9" href="#__codelineno-21-9"></a>âœ… Completed STRING enrichment results:
</span><span id="__span-21-10"><a id="__codelineno-21-10" name="__codelineno-21-10" href="#__codelineno-21-10"></a>    (none)
</span><span id="__span-21-11"><a id="__codelineno-21-11" name="__codelineno-21-11" href="#__codelineno-21-11"></a>
</span><span id="__span-21-12"><a id="__codelineno-21-12" name="__codelineno-21-12" href="#__codelineno-21-12"></a>âœ… Completed STRING PPI results:
</span><span id="__span-21-13"><a id="__codelineno-21-13" name="__codelineno-21-13" href="#__codelineno-21-13"></a>    (none)
</span></code></pre></div>
</div>
<p>Since we just ran a DE analysis, the key <strong><code>BE_kd vs BE_sc</code></strong> is available. We can run STRING <strong>functional enrichment</strong> on both up- and downregulated proteins.</p>
<div class="language-py highlight"><span class="filename">STRING functional enrichment</span><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a><span class="n">pdata</span><span class="o">.</span><span class="n">enrichment_functional</span><span class="p">(</span><span class="n">from_de</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">de_key</span><span class="o">=</span><span class="s2">&quot;BE_kd vs BE_sc&quot;</span><span class="p">)</span>
</span></code></pre></div>
<div class="result">
<div class="language-text highlight"><span class="filename">output</span><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>ðŸ§­ [USER] Running STRING enrichment [DE-based: [{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;kd&#39;}] vs [{&#39;cellline&#39;: &#39;BE&#39;, &#39;condition&#39;: &#39;sc&#39;}]]
</span><span id="__span-23-2"><a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a>
</span><span id="__span-23-3"><a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a>ðŸ”¹ Up-regulated proteins
</span><span id="__span-23-4"><a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a>     â„¹ï¸ Found 0 cached STRING IDs. 150 need lookup.
</span><span id="__span-23-5"><a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a>          â„¹ï¸ Cached 149 STRING IDs from UniProt API xref_string.
</span><span id="__span-23-6"><a id="__codelineno-23-6" name="__codelineno-23-6" href="#__codelineno-23-6"></a>          âš ï¸ No STRING mappings returned from STRING API.
</span><span id="__span-23-7"><a id="__codelineno-23-7" name="__codelineno-23-7" href="#__codelineno-23-7"></a>   ðŸ”¸ Proteins: 150 â†’ STRING IDs: 149
</span><span id="__span-23-8"><a id="__codelineno-23-8" name="__codelineno-23-8" href="#__codelineno-23-8"></a>   ðŸ”¸ Species: 9606 | Background: None
</span><span id="__span-23-9"><a id="__codelineno-23-9" name="__codelineno-23-9" href="#__codelineno-23-9"></a>     âœ… [OK] Enrichment complete (3.92s)
</span><span id="__span-23-10"><a id="__codelineno-23-10" name="__codelineno-23-10" href="#__codelineno-23-10"></a>   â€¢ Access result: pdata.stats[&#39;functional&#39;][&quot;BE_kd vs BE_sc_up&quot;][&quot;result&quot;]
</span><span id="__span-23-11"><a id="__codelineno-23-11" name="__codelineno-23-11" href="#__codelineno-23-11"></a>   â€¢ Plot command : pdata.plot_enrichment_svg(&quot;BE_kd vs BE_sc&quot;, direction=&quot;up&quot;)
</span><span id="__span-23-12"><a id="__codelineno-23-12" name="__codelineno-23-12" href="#__codelineno-23-12"></a>   â€¢ View online  : https://string-db.org/cgi/network?identifiers=9606.ENSP00000351310%0d9606....ENSP00000265018&amp;caller_identity=scpviz&amp;species=9606&amp;show_query_node_labels=1
</span><span id="__span-23-13"><a id="__codelineno-23-13" name="__codelineno-23-13" href="#__codelineno-23-13"></a>
</span><span id="__span-23-14"><a id="__codelineno-23-14" name="__codelineno-23-14" href="#__codelineno-23-14"></a>
</span><span id="__span-23-15"><a id="__codelineno-23-15" name="__codelineno-23-15" href="#__codelineno-23-15"></a>ðŸ”¹ Down-regulated proteins
</span><span id="__span-23-16"><a id="__codelineno-23-16" name="__codelineno-23-16" href="#__codelineno-23-16"></a>     â„¹ï¸ Found 0 cached STRING IDs. 150 need lookup.
</span><span id="__span-23-17"><a id="__codelineno-23-17" name="__codelineno-23-17" href="#__codelineno-23-17"></a>          â„¹ï¸ Cached 149 STRING IDs from UniProt API xref_string.
</span><span id="__span-23-18"><a id="__codelineno-23-18" name="__codelineno-23-18" href="#__codelineno-23-18"></a>          âš ï¸ No STRING mappings returned from STRING API.
</span><span id="__span-23-19"><a id="__codelineno-23-19" name="__codelineno-23-19" href="#__codelineno-23-19"></a>   ðŸ”¸ Proteins: 150 â†’ STRING IDs: 149
</span><span id="__span-23-20"><a id="__codelineno-23-20" name="__codelineno-23-20" href="#__codelineno-23-20"></a>   ðŸ”¸ Species: 9606 | Background: None
</span><span id="__span-23-21"><a id="__codelineno-23-21" name="__codelineno-23-21" href="#__codelineno-23-21"></a>     âœ… [OK] Enrichment complete (2.35s)
</span><span id="__span-23-22"><a id="__codelineno-23-22" name="__codelineno-23-22" href="#__codelineno-23-22"></a>   â€¢ Access result: pdata.stats[&#39;functional&#39;][&quot;BE_kd vs BE_sc_down&quot;][&quot;result&quot;]
</span><span id="__span-23-23"><a id="__codelineno-23-23" name="__codelineno-23-23" href="#__codelineno-23-23"></a><span class="hll">   â€¢ Plot command : pdata.plot_enrichment_svg(&quot;BE_kd vs BE_sc&quot;, direction=&quot;down&quot;)
</span></span><span id="__span-23-24"><a id="__codelineno-23-24" name="__codelineno-23-24" href="#__codelineno-23-24"></a>   â€¢ View online  : https://string-db.org/cgi/network?identifiers=9606.ENSP00000368678%0d9606.....ENSP00000263512%0d9606.ENSP00000382767%0d9606&amp;caller_identity=scpviz&amp;species=9606&amp;show_query_node_labels=1
</span></code></pre></div>
</div>
<p>Once enrichment is complete, you can visualize the <strong>Gene Ontology (Biological Process)</strong> results:</p>
<div class="language-py highlight"><span class="filename">Plot sample CVs</span><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a><span class="n">pdata</span><span class="o">.</span><span class="n">plot_enrichment_svg</span><span class="p">(</span><span class="s2">&quot;BE_kd vs BE_sc&quot;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s2">&quot;down&quot;</span><span class="p">)</span>
</span></code></pre></div>
<div class="result">
<figure>
  <img alt="Sample CV plot" src="../../assets/quickstart_6.svg" />
<br />
<figcaption></figcaption>
</figure>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Refer to the <a href="../enrichment/">Enrichment</a> tutorial for details on additional STRING features such as <strong>PPI networks</strong>, <strong>GSEA</strong>, and <strong>combined enrichment-PPI analyses</strong>.</p>
</div>
<h2 id="next-steps">Next steps</h2>
<p>For a complete workflow â€” from importing data to enrichment and network analysis â€” see the <a href="../">Tutorial Index</a>.</p>
<ul>
<li>Learn more about <strong>data import</strong> in <a href="../importing/">Importing Data</a>.  </li>
<li>Explore <strong>filtering options</strong> in <a href="../filtering/">Filtering</a>.  </li>
<li>Explore normalization and imputation options in <a href="../imputation/">Normalization &amp; Imputation</a>.  </li>
<li>Learn about differential expression and volcano plots in <a href="../de/">Differential Expression</a>.  </li>
<li>Perform functional and PPI enrichment in <a href="../enrichment/">STRING Enrichment</a>.  </li>
<li>See advanced visualization techniques in the <a href="../plotting/">Plotting Tutorial</a>.</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../installation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Installation">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Installation
              </div>
            </div>
          </a>
        
        
          
          <a href="../single_cell/" class="md-footer__link md-footer__link--next" aria-label="Next: Single-cell Example">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Single-cell Example
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2025 Marion Pang
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gnaprs" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://linkedin.com/in/marionpang" target="_blank" rel="noopener" title="linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3M135.4 416H69V202.2h66.5V416zM102.2 96a38.5 38.5 0 1 1 0 77 38.5 38.5 0 1 1 0-77m282.1 320h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://pypi.org/project/scpviz/" target="_blank" rel="noopener" title="pypi.org" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["navigation.footer", "content.footnote.tooltips", "content.code.copy", "content.code.select", "content.code.annotate", "search.highlight", "navigation.sections", "navigation.tabs", "navigation.path", "navigation.indexes"], "search": "../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../../javascripts/tablesort.js"></script>
      
    
  </body>
</html>