{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0a5080d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\srpang\\anaconda3\\envs\\py311\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from scpviz import pAnnData as pAnnData\n",
    "from scpviz import plotting as scplt\n",
    "from scpviz import utils as scutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "faf751a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Importing data of type [pd]\n",
      "--------------------------\n",
      "Starting import [Proteome Discoverer]\n",
      "\n",
      "Source file: ../assets/pd32_Proteins.txt / ../assets/pd32_PeptideSequenceGroups.txt\n",
      "Number of files: 12\n",
      "Proteins: 10393\n",
      "Peptides: 167114\n",
      "\n",
      "‚ÑπÔ∏è Using 'Modifications in Master Proteins' for modification annotation.\n",
      "     ‚ö†Ô∏è [WARN] Master proteins in the peptide matrix do not match proteins in the protein data, please check if files correspond to the same data.\n",
      "     ‚ÑπÔ∏è [INFO] If using PD3.2, this is a known issue due to changed protein grouping rules.\n",
      "\n",
      "‚ÑπÔ∏è 73 proteins with missing gene names.\n",
      "     üåê [API] Querying UniProt for batch 1/1 (73 proteins) [fields: accession, gene_primary]\n",
      "     ‚úÖ Retrieved UniProt metadata for 71 entries.\n",
      "     ‚úÖ [OK] Recovered 70 gene name(s) from UniProt. Genes found:\n",
      "         PCM1, HDLBP, GBF1, WDR36, DDX27, MTCL2, ATAD3A, SLC4A1AP, COG5, DBT...\n",
      "     ‚ö†Ô∏è [WARN] 3 gene name(s) still missing. Assigned as 'UNKNOWN_<accession>' for:\n",
      "         A0A0B4J2D5, Q6ZSR9, A9Z1Z3\n",
      "     üí° Tip: You can update these using `pdata.update_identifier_maps({'GENE': 'ACCESSION'}, on='protein', direction='reverse', overwrite=True)`\n",
      "\n",
      "     ‚ÑπÔ∏è [INFO] Using global q-values for 'prot' significance annotation.\n",
      "     ‚ÑπÔ∏è [INFO] Using global q-values for 'pep' significance annotation.\n",
      "\n",
      "‚ÑπÔ∏è RS matrix: (10393, 167114) (proteins √ó peptides), sparsity: 99.99%\n",
      "   - Proteins with ‚â•2 *unique* linked peptides: 10068/10393\n",
      "   - Peptides linked to ‚â•2 proteins: 5034/167114\n",
      "   - Mean peptides per protein: 16.59\n",
      "   - Mean proteins per peptide: 1.03\n",
      "     ‚úÖ [OK] pAnnData object is valid.\n",
      "     ‚úÖ [OK] Import complete. Use `print(pdata)` to view the object.\n",
      "--------------------------\n"
     ]
    }
   ],
   "source": [
    "obs_columns = ['Sample','cellline','treatment','condition','duration']\n",
    "\n",
    "pdata = pAnnData.import_data(source_type='pd', prot_file = '../assets/pd32_Proteins.txt', pep_file='../assets/pd32_PeptideSequenceGroups.txt', obs_columns=obs_columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1890383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "      <th>duration</th>\n",
       "      <th>protein_quant</th>\n",
       "      <th>protein_count</th>\n",
       "      <th>protein_abundance_sum</th>\n",
       "      <th>mbr_count</th>\n",
       "      <th>high_count</th>\n",
       "      <th>peptide_quant</th>\n",
       "      <th>peptide_count</th>\n",
       "      <th>peptide_abundance_sum</th>\n",
       "      <th>unique_pep2_protein_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F4</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.929472</td>\n",
       "      <td>9660</td>\n",
       "      <td>8.030133e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9653</td>\n",
       "      <td>0.836513</td>\n",
       "      <td>139793</td>\n",
       "      <td>8.029744e+10</td>\n",
       "      <td>9460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F5</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.925238</td>\n",
       "      <td>9616</td>\n",
       "      <td>1.027993e+11</td>\n",
       "      <td>0</td>\n",
       "      <td>9619</td>\n",
       "      <td>0.840360</td>\n",
       "      <td>140436</td>\n",
       "      <td>1.028033e+11</td>\n",
       "      <td>9420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F6</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.927836</td>\n",
       "      <td>9643</td>\n",
       "      <td>9.066725e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9638</td>\n",
       "      <td>0.807646</td>\n",
       "      <td>134969</td>\n",
       "      <td>9.068236e+10</td>\n",
       "      <td>9443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F10</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.904166</td>\n",
       "      <td>9397</td>\n",
       "      <td>4.078257e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9396</td>\n",
       "      <td>0.641335</td>\n",
       "      <td>107176</td>\n",
       "      <td>4.081629e+10</td>\n",
       "      <td>9238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F11</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.920620</td>\n",
       "      <td>9568</td>\n",
       "      <td>5.843762e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9581</td>\n",
       "      <td>0.710084</td>\n",
       "      <td>118665</td>\n",
       "      <td>5.857828e+10</td>\n",
       "      <td>9389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F12</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.881459</td>\n",
       "      <td>9161</td>\n",
       "      <td>2.939409e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9167</td>\n",
       "      <td>0.562113</td>\n",
       "      <td>93937</td>\n",
       "      <td>2.941074e+10</td>\n",
       "      <td>9022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F16</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.947465</td>\n",
       "      <td>9847</td>\n",
       "      <td>9.022711e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9861</td>\n",
       "      <td>0.844561</td>\n",
       "      <td>141138</td>\n",
       "      <td>9.026093e+10</td>\n",
       "      <td>9643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F17</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.942269</td>\n",
       "      <td>9793</td>\n",
       "      <td>8.304290e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9783</td>\n",
       "      <td>0.793051</td>\n",
       "      <td>132530</td>\n",
       "      <td>8.314867e+10</td>\n",
       "      <td>9591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F18</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.948331</td>\n",
       "      <td>9856</td>\n",
       "      <td>8.971543e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9864</td>\n",
       "      <td>0.844783</td>\n",
       "      <td>141175</td>\n",
       "      <td>8.973545e+10</td>\n",
       "      <td>9641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F22</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.942942</td>\n",
       "      <td>9800</td>\n",
       "      <td>7.545933e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9806</td>\n",
       "      <td>0.747891</td>\n",
       "      <td>124983</td>\n",
       "      <td>7.554042e+10</td>\n",
       "      <td>9619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F23</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.165881</td>\n",
       "      <td>1724</td>\n",
       "      <td>1.037960e+08</td>\n",
       "      <td>0</td>\n",
       "      <td>1667</td>\n",
       "      <td>0.027131</td>\n",
       "      <td>4534</td>\n",
       "      <td>1.040278e+08</td>\n",
       "      <td>1712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F24</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.937939</td>\n",
       "      <td>9748</td>\n",
       "      <td>7.946756e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9757</td>\n",
       "      <td>0.799903</td>\n",
       "      <td>133675</td>\n",
       "      <td>7.951834e+10</td>\n",
       "      <td>9552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample cellline treatment condition duration  protein_quant  \\\n",
       "F4   Sample       AS        RA        kd       d7       0.929472   \n",
       "F5   Sample       AS        RA        kd       d7       0.925238   \n",
       "F6   Sample       AS        RA        kd       d7       0.927836   \n",
       "F10  Sample       AS        RA        sc       d7       0.904166   \n",
       "F11  Sample       AS        RA        sc       d7       0.920620   \n",
       "F12  Sample       AS        RA        sc       d7       0.881459   \n",
       "F16  Sample       BE        RA        kd       d7       0.947465   \n",
       "F17  Sample       BE        RA        kd       d7       0.942269   \n",
       "F18  Sample       BE        RA        kd       d7       0.948331   \n",
       "F22  Sample       BE        RA        sc       d7       0.942942   \n",
       "F23  Sample       BE        RA        sc       d7       0.165881   \n",
       "F24  Sample       BE        RA        sc       d7       0.937939   \n",
       "\n",
       "     protein_count  protein_abundance_sum  mbr_count  high_count  \\\n",
       "F4            9660           8.030133e+10          0        9653   \n",
       "F5            9616           1.027993e+11          0        9619   \n",
       "F6            9643           9.066725e+10          0        9638   \n",
       "F10           9397           4.078257e+10          0        9396   \n",
       "F11           9568           5.843762e+10          0        9581   \n",
       "F12           9161           2.939409e+10          0        9167   \n",
       "F16           9847           9.022711e+10          0        9861   \n",
       "F17           9793           8.304290e+10          0        9783   \n",
       "F18           9856           8.971543e+10          0        9864   \n",
       "F22           9800           7.545933e+10          0        9806   \n",
       "F23           1724           1.037960e+08          0        1667   \n",
       "F24           9748           7.946756e+10          0        9757   \n",
       "\n",
       "     peptide_quant  peptide_count  peptide_abundance_sum  \\\n",
       "F4        0.836513         139793           8.029744e+10   \n",
       "F5        0.840360         140436           1.028033e+11   \n",
       "F6        0.807646         134969           9.068236e+10   \n",
       "F10       0.641335         107176           4.081629e+10   \n",
       "F11       0.710084         118665           5.857828e+10   \n",
       "F12       0.562113          93937           2.941074e+10   \n",
       "F16       0.844561         141138           9.026093e+10   \n",
       "F17       0.793051         132530           8.314867e+10   \n",
       "F18       0.844783         141175           8.973545e+10   \n",
       "F22       0.747891         124983           7.554042e+10   \n",
       "F23       0.027131           4534           1.040278e+08   \n",
       "F24       0.799903         133675           7.951834e+10   \n",
       "\n",
       "     unique_pep2_protein_count  \n",
       "F4                        9460  \n",
       "F5                        9420  \n",
       "F6                        9443  \n",
       "F10                       9238  \n",
       "F11                       9389  \n",
       "F12                       9022  \n",
       "F16                       9643  \n",
       "F17                       9591  \n",
       "F18                       9641  \n",
       "F22                       9619  \n",
       "F23                       1712  \n",
       "F24                       9552  \n",
       "‚ö†Ô∏è  [TrackedDataFrame] This summary has been modified and is not synced back to .obs."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata.summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d73c2c",
   "metadata": {},
   "source": [
    "# Tutorial 2: Filtering and Normalization\n",
    "\n",
    "Learn how to filter proteins and peptides in your dataset.\n",
    "\n",
    "All filter functions return a **copy** of the `pAnnData` object unless `inplace=True` is specified.\n",
    "\n",
    "There are three main filtering functions:\n",
    "\n",
    "* `pdata.filter_prot()` ‚Äì filters proteins from the dataset  \n",
    "    * `pdata.filter_prot_found()` ‚Äì sub-function for proteins found within specified samples  \n",
    "    * `pdata.filter_prot_significant()` ‚Äì sub-function for proteins significant within specified samples  \n",
    "* `pdata.filter_sample()` ‚Äì filters samples from the dataset  \n",
    "* `pdata.filter_rs()` ‚Äì RS-based filtering (e.g., filtering by unique peptides)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00259a1a",
   "metadata": {},
   "source": [
    "## `filter_prot()`\n",
    "\n",
    "Filter protein data based on metadata conditions or accession lists (protein or gene names)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e22257c",
   "metadata": {},
   "source": [
    "### Condition-based filtering\n",
    "A condition string to filter protein metadata. Supports:\n",
    "* Standard comparisons, e.g. `\"Protein FDR Confidence: Combined == 'High'\"`\n",
    "* Substring queries using `includes`, e.g. `\"Description includes 'p97'\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "42c831d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [condition]:\n",
      "    Returning a copy of protein data based on condition:\n",
      "    ‚Üí condition: Protein FDR Confidence: Combined == 'High'\n",
      "    ‚Üí Proteins kept: 10393\n",
      "    ‚Üí Peptides kept (linked): 165954\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Filter by metadata condition\n",
    "condition = \"Protein FDR Confidence: Combined == 'High'\"\n",
    "pdata_filtered = pdata.filter_prot(condition=condition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "318f6d21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [condition]:\n",
      "    Returning a copy of protein data based on condition:\n",
      "    ‚Üí condition: Description includes 'VCP'\n",
      "    ‚Üí Proteins kept: 4\n",
      "    ‚Üí Peptides kept (linked): 116\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Checked",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Protein Group FDR Confidence: Combined",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Protein FDR Confidence: Combined",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Master",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Description",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Exp. Protein Group q-value: Combined",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Exp. Protein q-value: Combined",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Sum PEP Score",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Coverage [%]",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "# Peptides",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "# PSMs",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "# Protein Unique Peptides",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "unique_peptides",
         "rawType": "int32",
         "type": "integer"
        },
        {
         "name": "# AAs",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "MW [kDa]",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "calc. pI",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Score CHIMERYS on Ardia Server: CHIMERYS on Ardia Server",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "# Peptides (by Search Engine): CHIMERYS on Ardia Server",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Biological Process",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Cellular Component",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Molecular Function",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Pfam IDs",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Entrez Gene ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Ensembl Gene ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Genes",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Gene ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Reactome Pathways",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "# Protein Pathway Groups",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "# Razor Peptides",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Found in Sample: F4: Sample, AS, RA, kd, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F5: Sample, AS, RA, kd, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F6: Sample, AS, RA, kd, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F10: Sample, AS, RA, sc, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F11: Sample, AS, RA, sc, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F12: Sample, AS, RA, sc, d7",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "Found in Sample: F16: Sample, BE, RA, kd, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F17: Sample, BE, RA, kd, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F18: Sample, BE, RA, kd, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F22: Sample, BE, RA, sc, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Found in Sample: F23: Sample, BE, RA, sc, d7",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "Found in Sample: F24: Sample, BE, RA, sc, d7",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "# Protein Groups",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Modifications",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Global_Q_value",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "peptides_per_protein",
         "rawType": "int32",
         "type": "integer"
        },
        {
         "name": "Found In: F4",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F5",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F6",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F10",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F11",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F12",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F16",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F17",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F18",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F22",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F23",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Found In: F24",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "Significant In: Global",
         "rawType": "bool",
         "type": "boolean"
        }
       ],
       "ref": "b5fe2acf-e34b-4dfd-9ec4-09d4f28e3bc2",
       "rows": [
        [
         "P55072",
         "False",
         "High",
         "High",
         "IsMasterProtein",
         "P55072",
         "Transitional endoplasmic reticulum ATPase OS=Homo sapiens OX=9606 GN=VCP PE=1 SV=4",
         "0.0",
         "0.0",
         "292.503",
         "70",
         "58",
         "1000",
         "0",
         "58",
         "806",
         "89.3",
         "5.26",
         "454.82",
         "58",
         "cell cycle OR cell proliferation;cell organization and biogenesis;protein metabolism;DNA metabolism;other metabolic processes;stress response;transport;signal transduction;other biological processes",
         "non-structural extracellular;other membranes;cytosol;cytoskeleton;ER/Golgi;nucleus;other cytoplasmic organelle;other cell component",
         "signal transduction activity or receptor binding;enzyme regulator activity;other molecular function",
         "Pf00004, Pf02359, Pf02933, Pf17862",
         "7415",
         "ENSG00000165280",
         "VCP",
         "B2R5T8; hsa:7415; P55072; Q0V924; Q2TAI5; Q969G7; Q9UCD5; V9HW80",
         "Neutrophil degranulation; RHOH GTPase cycle; E3 ubiquitin ligases ubiquitinate target proteins; Aggrephagy; Neddylation; ABC-family proteins mediated transport; Hedgehog ligand biogenesis; Hh mutants are degraded by ERAD; Defective CFTR causes cystic fibrosis; KEAP1-NFE2L2 pathway; AMPK-induced ERAD and lysosome mediated degradation of PD-L1(CD274); Ribosome Quality Control (RQC) complex extracts and degrades nascent peptide; Attachment and Entry; Attachment and Entry; Ovarian tumor domain proteases; Josephin domain DUBs; Protein methylation; HSF1 activation; Translesion Synthesis by POLH; N-glycan trimming in the ER and Calnexin/Calreticulin cycle",
         "23",
         "0",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "1",
         null,
         "0.0",
         "58",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True"
        ],
        [
         "Q96JH7",
         "False",
         "High",
         "High",
         "IsMasterProtein",
         "Q96JH7",
         "Deubiquitinating protein VCPIP1 OS=Homo sapiens OX=9606 GN=VCPIP1 PE=1 SV=2",
         "0.0",
         "0.0",
         "180.522",
         "54",
         "53",
         "662",
         "0",
         "53",
         "1222",
         "134.2",
         "7.2",
         "202.92",
         "53",
         "cell organization and biogenesis;protein metabolism;DNA metabolism;other metabolic processes;stress response;other biological processes",
         "ER/Golgi;nucleus;other cell component",
         "other molecular function",
         "Pf02338, Pf19437, Pf21403",
         "80124",
         "ENSG00000175073",
         "VCPIP1",
         "hsa:80124; Q504T4; Q86T93; Q86W01; Q8N3A9; Q96JH7; Q9H5R8",
         "Ovarian tumor domain proteases",
         "2",
         "0",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "1",
         null,
         "0.0",
         "53",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True"
        ],
        [
         "Q8NHG7",
         "False",
         "High",
         "High",
         "IsMasterProtein",
         "Q8NHG7",
         "Small VCP/p97-interacting protein OS=Homo sapiens OX=9606 GN=SVIP PE=1 SV=1",
         "0.00033",
         "0.00031",
         "8.631",
         "32",
         "2",
         "22",
         "0",
         "2",
         "77",
         "8.4",
         "8.91",
         "8.02",
         "2",
         "other biological processes",
         "plasma membrane;other membranes;ER/Golgi;other cytoplasmic organelle",
         "other molecular function",
         "Pf15811",
         "258010",
         "ENSG00000198168",
         "SVIP",
         "hsa:258010; Q8NHG7",
         "Neutrophil degranulation",
         "1",
         "0",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         "High",
         null,
         "High",
         "1",
         null,
         "0.00033",
         "2",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True"
        ],
        [
         "Q9H867",
         "False",
         "High",
         "High",
         "IsMasterProtein",
         "Q9H867",
         "Protein N-lysine methyltransferase METTL21D OS=Homo sapiens OX=9606 GN=VCPKMT PE=1 SV=2",
         "0.00084",
         "0.0008",
         "6.936",
         "16",
         "3",
         "24",
         "0",
         "3",
         "229",
         "25.8",
         "4.92",
         "4.52",
         "3",
         "protein metabolism;other metabolic processes;other biological processes",
         "cytosol;other cell component",
         "other molecular function",
         "Pf10294",
         "79609",
         "ENSG00000100483",
         "VCPKMT",
         "B7ZLA3; B7ZLA4; hsa:79609; Q2M2X3; Q86T12; Q9H867",
         "Protein methylation",
         "1",
         "0",
         "High",
         "High",
         "High",
         "High",
         "High",
         null,
         "High",
         "High",
         "High",
         "High",
         null,
         "High",
         "1",
         null,
         "0.00084",
         "3",
         "True",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True"
        ]
       ],
       "shape": {
        "columns": 59,
        "rows": 4
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked</th>\n",
       "      <th>Protein Group FDR Confidence: Combined</th>\n",
       "      <th>Protein FDR Confidence: Combined</th>\n",
       "      <th>Master</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Exp. Protein Group q-value: Combined</th>\n",
       "      <th>Exp. Protein q-value: Combined</th>\n",
       "      <th>Sum PEP Score</th>\n",
       "      <th>Coverage [%]</th>\n",
       "      <th>...</th>\n",
       "      <th>Found In: F10</th>\n",
       "      <th>Found In: F11</th>\n",
       "      <th>Found In: F12</th>\n",
       "      <th>Found In: F16</th>\n",
       "      <th>Found In: F17</th>\n",
       "      <th>Found In: F18</th>\n",
       "      <th>Found In: F22</th>\n",
       "      <th>Found In: F23</th>\n",
       "      <th>Found In: F24</th>\n",
       "      <th>Significant In: Global</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P55072</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>IsMasterProtein</td>\n",
       "      <td>P55072</td>\n",
       "      <td>Transitional endoplasmic reticulum ATPase OS=H...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>292.503</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q96JH7</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>IsMasterProtein</td>\n",
       "      <td>Q96JH7</td>\n",
       "      <td>Deubiquitinating protein VCPIP1 OS=Homo sapien...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>180.522</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q8NHG7</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>IsMasterProtein</td>\n",
       "      <td>Q8NHG7</td>\n",
       "      <td>Small VCP/p97-interacting protein OS=Homo sapi...</td>\n",
       "      <td>0.00033</td>\n",
       "      <td>0.00031</td>\n",
       "      <td>8.631</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9H867</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>IsMasterProtein</td>\n",
       "      <td>Q9H867</td>\n",
       "      <td>Protein N-lysine methyltransferase METTL21D OS...</td>\n",
       "      <td>0.00084</td>\n",
       "      <td>0.00080</td>\n",
       "      <td>6.936</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows √ó 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Checked Protein Group FDR Confidence: Combined  \\\n",
       "P55072    False                                   High   \n",
       "Q96JH7    False                                   High   \n",
       "Q8NHG7    False                                   High   \n",
       "Q9H867    False                                   High   \n",
       "\n",
       "       Protein FDR Confidence: Combined           Master Accession  \\\n",
       "P55072                             High  IsMasterProtein    P55072   \n",
       "Q96JH7                             High  IsMasterProtein    Q96JH7   \n",
       "Q8NHG7                             High  IsMasterProtein    Q8NHG7   \n",
       "Q9H867                             High  IsMasterProtein    Q9H867   \n",
       "\n",
       "                                              Description  \\\n",
       "P55072  Transitional endoplasmic reticulum ATPase OS=H...   \n",
       "Q96JH7  Deubiquitinating protein VCPIP1 OS=Homo sapien...   \n",
       "Q8NHG7  Small VCP/p97-interacting protein OS=Homo sapi...   \n",
       "Q9H867  Protein N-lysine methyltransferase METTL21D OS...   \n",
       "\n",
       "        Exp. Protein Group q-value: Combined  Exp. Protein q-value: Combined  \\\n",
       "P55072                               0.00000                         0.00000   \n",
       "Q96JH7                               0.00000                         0.00000   \n",
       "Q8NHG7                               0.00033                         0.00031   \n",
       "Q9H867                               0.00084                         0.00080   \n",
       "\n",
       "        Sum PEP Score  Coverage [%]  ...  Found In: F10  Found In: F11  \\\n",
       "P55072        292.503            70  ...           True           True   \n",
       "Q96JH7        180.522            54  ...           True           True   \n",
       "Q8NHG7          8.631            32  ...           True           True   \n",
       "Q9H867          6.936            16  ...           True           True   \n",
       "\n",
       "        Found In: F12  Found In: F16  Found In: F17  Found In: F18  \\\n",
       "P55072           True           True           True           True   \n",
       "Q96JH7           True           True           True           True   \n",
       "Q8NHG7           True           True           True           True   \n",
       "Q9H867          False           True           True           True   \n",
       "\n",
       "        Found In: F22  Found In: F23  Found In: F24 Significant In: Global  \n",
       "P55072           True           True           True                   True  \n",
       "Q96JH7           True           True           True                   True  \n",
       "Q8NHG7           True          False           True                   True  \n",
       "Q9H867           True          False           True                   True  \n",
       "\n",
       "[4 rows x 59 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Substring match on protein description\n",
    "condition = \"Description includes 'VCP'\"\n",
    "pdata_filtered = pdata.filter_prot(condition=condition)\n",
    "pdata_filtered.prot.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e0ff5d5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [condition]:\n",
      "    Returning a copy of protein data based on condition:\n",
      "    ‚Üí condition: unique_peptides >= 2\n",
      "    ‚Üí Proteins kept: 10068\n",
      "    ‚Üí Peptides kept (linked): 165468\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Numerical condition on metadata\n",
    "pdata_filtered = pdata.filter_prot(condition=\"unique_peptides >= 2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "129b8940",
   "metadata": {},
   "source": [
    "### Accession-based filtering (accession list or gene names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1a6f500a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [accession]:\n",
      "    Returning a copy of protein data based on accession:\n",
      "    ‚Üí accessions: 1 matched / 2 requested\n",
      "    ‚Üí Proteins kept: 1\n",
      "    ‚Üí Peptides kept (linked): 20\n",
      "\n"
     ]
    }
   ],
   "source": [
    "accessions = ['GAPDH', 'P53']\n",
    "pdata_filtered = pdata.filter_prot(accessions=accessions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b11342",
   "metadata": {},
   "source": [
    "### Valid genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "be1e7b50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [valid genes]:\n",
      "    Returning a copy of protein data based on valid genes:\n",
      "    ‚Üí valid_genes: removed 0 proteins with invalid gene names\n",
      "    ‚Üí valid_genes: resolved 12 duplicate gene names by appending numeric suffixes\n",
      "    ‚Üí Proteins kept: 10393\n",
      "    ‚Üí Peptides kept (linked): 165954\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Filter proteins with valid genes only\n",
    "pdata_filtered = pdata.filter_prot(valid_genes=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b0ee7dc",
   "metadata": {},
   "source": [
    "### Unique profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "20bda755",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [unique profiles]:\n",
      "    Returning a copy of protein data based on unique profiles:\n",
      "    ‚Üí unique_profiles: removed 0 duplicate and 0 empty abundance profiles (0 total)\n",
      "    ‚Üí Proteins kept: 10393\n",
      "    ‚Üí Peptides kept (linked): 165954\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Filter proteins with unique abundance profiles\n",
    "pdata_filtered = pdata.filter_prot(unique_profiles=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37557973",
   "metadata": {},
   "source": [
    "## `filter_prot_found()`\n",
    "\n",
    "Filter proteins or peptides based on *\"Found In\"* detection across samples or groups.\n",
    "\n",
    "This method filters features by checking whether they are found in a minimum number or proportion of samples, either at the group level (e.g., biological condition) or based on individual files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "31af12d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     ‚ÑπÔ∏è [INFO] Found matching obs column(s): ['kd', 'sc']. Automatically annotating detection by group values.\n",
      "üß≠ [USER] Filtering proteins [Found|Group-mode|ALL]: keeping 10001 / 10393 features passing threshold 2 across groups: ['kd', 'sc']\n",
      "     ‚ÑπÔ∏è [INFO] Applying RS-based peptide sync-up on peptides after protein filtering...\n",
      "     ‚úÖ [OK] RS matrix filtered: 10001 proteins, 164622 peptides retained.\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_prot_found(group=[\"condition\"], min_count=2, match_any=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "46858649",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     ‚ÑπÔ∏è [INFO] Found matching obs column(s): ['kd', 'sc']. Automatically annotating detection by group values.\n",
      "üß≠ [USER] Filtering proteins [Found|Group-mode|ANY]: keeping 10273 / 10393 features passing threshold in ANY of groups: ['kd', 'sc']\n",
      "     ‚ÑπÔ∏è [INFO] Applying RS-based peptide sync-up on peptides after protein filtering...\n",
      "     ‚úÖ [OK] RS matrix filtered: 10273 proteins, 165571 peptides retained.\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_prot_found(group=[\"condition\"], min_count=2, match_any=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aae7b7c",
   "metadata": {},
   "source": [
    "Note that with the weaker OR condition in `match_any=True`, more proteins pass the filter!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05e65918",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[summary] ‚ö†Ô∏è Warning: .summary has been modified. Run `(pdata).update_summary()` to sync changes back to .obs.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "      <th>duration</th>\n",
       "      <th>protein_quant</th>\n",
       "      <th>protein_count</th>\n",
       "      <th>protein_abundance_sum</th>\n",
       "      <th>mbr_count</th>\n",
       "      <th>high_count</th>\n",
       "      <th>peptide_quant</th>\n",
       "      <th>peptide_count</th>\n",
       "      <th>peptide_abundance_sum</th>\n",
       "      <th>unique_pep2_protein_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F4</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.929472</td>\n",
       "      <td>9660</td>\n",
       "      <td>8.030133e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9653</td>\n",
       "      <td>0.836513</td>\n",
       "      <td>139793</td>\n",
       "      <td>8.029744e+10</td>\n",
       "      <td>9460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F5</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.925238</td>\n",
       "      <td>9616</td>\n",
       "      <td>1.027993e+11</td>\n",
       "      <td>0</td>\n",
       "      <td>9619</td>\n",
       "      <td>0.840360</td>\n",
       "      <td>140436</td>\n",
       "      <td>1.028033e+11</td>\n",
       "      <td>9420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F6</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.927836</td>\n",
       "      <td>9643</td>\n",
       "      <td>9.066725e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9638</td>\n",
       "      <td>0.807646</td>\n",
       "      <td>134969</td>\n",
       "      <td>9.068236e+10</td>\n",
       "      <td>9443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F10</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.904166</td>\n",
       "      <td>9397</td>\n",
       "      <td>4.078257e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9396</td>\n",
       "      <td>0.641335</td>\n",
       "      <td>107176</td>\n",
       "      <td>4.081629e+10</td>\n",
       "      <td>9238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F11</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.920620</td>\n",
       "      <td>9568</td>\n",
       "      <td>5.843762e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9581</td>\n",
       "      <td>0.710084</td>\n",
       "      <td>118665</td>\n",
       "      <td>5.857828e+10</td>\n",
       "      <td>9389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F12</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.881459</td>\n",
       "      <td>9161</td>\n",
       "      <td>2.939409e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9167</td>\n",
       "      <td>0.562113</td>\n",
       "      <td>93937</td>\n",
       "      <td>2.941074e+10</td>\n",
       "      <td>9022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F16</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.947465</td>\n",
       "      <td>9847</td>\n",
       "      <td>9.022711e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9861</td>\n",
       "      <td>0.844561</td>\n",
       "      <td>141138</td>\n",
       "      <td>9.026093e+10</td>\n",
       "      <td>9643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F17</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.942269</td>\n",
       "      <td>9793</td>\n",
       "      <td>8.304290e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9783</td>\n",
       "      <td>0.793051</td>\n",
       "      <td>132530</td>\n",
       "      <td>8.314867e+10</td>\n",
       "      <td>9591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F18</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.948331</td>\n",
       "      <td>9856</td>\n",
       "      <td>8.971543e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9864</td>\n",
       "      <td>0.844783</td>\n",
       "      <td>141175</td>\n",
       "      <td>8.973545e+10</td>\n",
       "      <td>9641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F22</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.942942</td>\n",
       "      <td>9800</td>\n",
       "      <td>7.545933e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9806</td>\n",
       "      <td>0.747891</td>\n",
       "      <td>124983</td>\n",
       "      <td>7.554042e+10</td>\n",
       "      <td>9619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F23</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.165881</td>\n",
       "      <td>1724</td>\n",
       "      <td>1.037960e+08</td>\n",
       "      <td>0</td>\n",
       "      <td>1667</td>\n",
       "      <td>0.027131</td>\n",
       "      <td>4534</td>\n",
       "      <td>1.040278e+08</td>\n",
       "      <td>1712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F24</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.937939</td>\n",
       "      <td>9748</td>\n",
       "      <td>7.946756e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9757</td>\n",
       "      <td>0.799903</td>\n",
       "      <td>133675</td>\n",
       "      <td>7.951834e+10</td>\n",
       "      <td>9552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample cellline treatment condition duration  protein_quant  \\\n",
       "F4   Sample       AS        RA        kd       d7       0.929472   \n",
       "F5   Sample       AS        RA        kd       d7       0.925238   \n",
       "F6   Sample       AS        RA        kd       d7       0.927836   \n",
       "F10  Sample       AS        RA        sc       d7       0.904166   \n",
       "F11  Sample       AS        RA        sc       d7       0.920620   \n",
       "F12  Sample       AS        RA        sc       d7       0.881459   \n",
       "F16  Sample       BE        RA        kd       d7       0.947465   \n",
       "F17  Sample       BE        RA        kd       d7       0.942269   \n",
       "F18  Sample       BE        RA        kd       d7       0.948331   \n",
       "F22  Sample       BE        RA        sc       d7       0.942942   \n",
       "F23  Sample       BE        RA        sc       d7       0.165881   \n",
       "F24  Sample       BE        RA        sc       d7       0.937939   \n",
       "\n",
       "     protein_count  protein_abundance_sum  mbr_count  high_count  \\\n",
       "F4            9660           8.030133e+10          0        9653   \n",
       "F5            9616           1.027993e+11          0        9619   \n",
       "F6            9643           9.066725e+10          0        9638   \n",
       "F10           9397           4.078257e+10          0        9396   \n",
       "F11           9568           5.843762e+10          0        9581   \n",
       "F12           9161           2.939409e+10          0        9167   \n",
       "F16           9847           9.022711e+10          0        9861   \n",
       "F17           9793           8.304290e+10          0        9783   \n",
       "F18           9856           8.971543e+10          0        9864   \n",
       "F22           9800           7.545933e+10          0        9806   \n",
       "F23           1724           1.037960e+08          0        1667   \n",
       "F24           9748           7.946756e+10          0        9757   \n",
       "\n",
       "     peptide_quant  peptide_count  peptide_abundance_sum  \\\n",
       "F4        0.836513         139793           8.029744e+10   \n",
       "F5        0.840360         140436           1.028033e+11   \n",
       "F6        0.807646         134969           9.068236e+10   \n",
       "F10       0.641335         107176           4.081629e+10   \n",
       "F11       0.710084         118665           5.857828e+10   \n",
       "F12       0.562113          93937           2.941074e+10   \n",
       "F16       0.844561         141138           9.026093e+10   \n",
       "F17       0.793051         132530           8.314867e+10   \n",
       "F18       0.844783         141175           8.973545e+10   \n",
       "F22       0.747891         124983           7.554042e+10   \n",
       "F23       0.027131           4534           1.040278e+08   \n",
       "F24       0.799903         133675           7.951834e+10   \n",
       "\n",
       "     unique_pep2_protein_count  \n",
       "F4                        9460  \n",
       "F5                        9420  \n",
       "F6                        9443  \n",
       "F10                       9238  \n",
       "F11                       9389  \n",
       "F12                       9022  \n",
       "F16                       9643  \n",
       "F17                       9591  \n",
       "F18                       9641  \n",
       "F22                       9619  \n",
       "F23                       1712  \n",
       "F24                       9552  \n",
       "‚ö†Ô∏è  [TrackedDataFrame] This summary has been modified and is not synced back to .obs."
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata.summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7219d81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering proteins [Found|File-mode|ALL]: keeping 9379 / 10393 features found in ALL files: ['F4', 'F5', 'F6']\n",
      "     ‚ÑπÔ∏è [INFO] Applying RS-based peptide sync-up on peptides after protein filtering...\n",
      "     ‚úÖ [OK] RS matrix filtered: 9379 proteins, 160654 peptides retained.\n"
     ]
    }
   ],
   "source": [
    "# all the AS, RA, kd files\n",
    "pdata_filtered = pdata.filter_prot_found(group=[\"F4\", \"F5\", \"F6\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68fe0de5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Annotated 'found in' features by group: ['cellline', 'condition'] using threshold 0.0.\n",
      "üß≠ [USER] Filtering proteins [Found|Group-mode|ALL]: keeping 9147 / 10393 features passing threshold 0.5 across groups: ['AS_kd', 'BE_sc']\n",
      "     ‚ÑπÔ∏è [INFO] Applying RS-based peptide sync-up on peptides after protein filtering...\n",
      "     ‚úÖ [OK] RS matrix filtered: 9147 proteins, 159410 peptides retained.\n"
     ]
    }
   ],
   "source": [
    "# proteins found in both AS_kd and BE_sc\n",
    "pdata.annotate_found(classes=['cellline', 'condition'])\n",
    "pdata_filtered = pdata.filter_prot_found(group=['AS_kd', 'BE_sc'],min_ratio=0.5,match_any=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5034c119",
   "metadata": {},
   "source": [
    "## filter_prot_significant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f3289211",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     ‚ÑπÔ∏è [INFO] Applying RS-based peptide sync-up on peptides after protein filtering...\n",
      "     ‚úÖ [OK] RS matrix filtered: 10393 proteins, 165954 peptides retained.\n",
      "üß≠ [USER] Filtering proteins by significance [Global-mode]:\n",
      "     ‚ÑπÔ∏è [INFO] Using global protein-level q-values (no per-sample significance available).\n",
      "    Returning a copy of protein data based on significance thresholds:\n",
      "     üî∏ Files requested: All\n",
      "     üî∏ FDR threshold: 0.01\n",
      "    ‚Üí Proteins kept: 10393, Proteins dropped: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_prot_significant()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9dbb35f",
   "metadata": {},
   "source": [
    "## `filter_sample()`\n",
    "\n",
    "Filter samples in a `pAnnData` object based on categorical, numeric, or identifier-based criteria.  \n",
    "Accepts **exactly one** of the following arguments:\n",
    "\n",
    "- `values`: A dictionary or list of dictionaries specifying class-based filters (e.g., treatment, cellline).  \n",
    "- `condition`: A logical condition string evaluated against summary-level numeric metadata (e.g., protein count).  \n",
    "- `file_list`: A list of sample or file names to retain."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c0229f0",
   "metadata": {},
   "source": [
    "### Filter by value\n",
    "\n",
    "Categorical metadata filtering allows selection of samples based on `.obs` or `.summary` fields such as treatment, cell line, or condition.  \n",
    "This supports:\n",
    "\n",
    "- A **single dictionary**, e.g. `{'cellline': 'A'}`  \n",
    "- A **list of dictionaries** for multiple matching cases, e.g. `[{...}, {...}]`  \n",
    "- **Exact matching**: \n",
    "    * `exact_cases=True` for strict combination matching across all key‚Äìvalue pairs\n",
    "    * `exact_cases=False` applies an OR logic within fields and AND logic across fields.\n",
    "\n",
    "#### `exact_cases=False`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8f14b145",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering samples [class match]:\n",
      "    Returning a copy of sample data based on class match:\n",
      "   üî∏ Match samples where:\n",
      "      - condition: kd, sc\n",
      "      - cellline: AS\n",
      "     ‚ÑπÔ∏è Auto-cleanup: Removed 299 empty proteins (all-NaN or all-zero). Proteins: O15523, O00507, Q8TDI0, Q01814, Q14576, Q13585, O60641, P52895, Q9BY66, Q4G176, ...\n",
      "    ‚Üí Samples kept: 6, Samples dropped: 6\n",
      "    ‚Üí Proteins kept: 10094\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "      <th>duration</th>\n",
       "      <th>protein_quant</th>\n",
       "      <th>protein_count</th>\n",
       "      <th>protein_abundance_sum</th>\n",
       "      <th>mbr_count</th>\n",
       "      <th>high_count</th>\n",
       "      <th>peptide_quant</th>\n",
       "      <th>peptide_count</th>\n",
       "      <th>peptide_abundance_sum</th>\n",
       "      <th>unique_pep2_protein_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F4</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.957004</td>\n",
       "      <td>9660</td>\n",
       "      <td>8.030133e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9643</td>\n",
       "      <td>0.846334</td>\n",
       "      <td>139348</td>\n",
       "      <td>8.024645e+10</td>\n",
       "      <td>9462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F5</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.952645</td>\n",
       "      <td>9616</td>\n",
       "      <td>1.027993e+11</td>\n",
       "      <td>0</td>\n",
       "      <td>9608</td>\n",
       "      <td>0.850245</td>\n",
       "      <td>139992</td>\n",
       "      <td>1.027483e+11</td>\n",
       "      <td>9421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F6</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.955320</td>\n",
       "      <td>9643</td>\n",
       "      <td>9.066725e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9630</td>\n",
       "      <td>0.817193</td>\n",
       "      <td>134550</td>\n",
       "      <td>9.062305e+10</td>\n",
       "      <td>9445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F10</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.930949</td>\n",
       "      <td>9397</td>\n",
       "      <td>4.078257e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9387</td>\n",
       "      <td>0.648756</td>\n",
       "      <td>106817</td>\n",
       "      <td>4.076389e+10</td>\n",
       "      <td>9239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F11</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.947890</td>\n",
       "      <td>9568</td>\n",
       "      <td>5.843762e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9572</td>\n",
       "      <td>0.718109</td>\n",
       "      <td>118236</td>\n",
       "      <td>5.850732e+10</td>\n",
       "      <td>9391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F12</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.907569</td>\n",
       "      <td>9161</td>\n",
       "      <td>2.939409e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9161</td>\n",
       "      <td>0.568689</td>\n",
       "      <td>93634</td>\n",
       "      <td>2.937665e+10</td>\n",
       "      <td>9022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample cellline treatment condition duration  protein_quant  \\\n",
       "F4   Sample       AS        RA        kd       d7       0.957004   \n",
       "F5   Sample       AS        RA        kd       d7       0.952645   \n",
       "F6   Sample       AS        RA        kd       d7       0.955320   \n",
       "F10  Sample       AS        RA        sc       d7       0.930949   \n",
       "F11  Sample       AS        RA        sc       d7       0.947890   \n",
       "F12  Sample       AS        RA        sc       d7       0.907569   \n",
       "\n",
       "     protein_count  protein_abundance_sum  mbr_count  high_count  \\\n",
       "F4            9660           8.030133e+10          0        9643   \n",
       "F5            9616           1.027993e+11          0        9608   \n",
       "F6            9643           9.066725e+10          0        9630   \n",
       "F10           9397           4.078257e+10          0        9387   \n",
       "F11           9568           5.843762e+10          0        9572   \n",
       "F12           9161           2.939409e+10          0        9161   \n",
       "\n",
       "     peptide_quant  peptide_count  peptide_abundance_sum  \\\n",
       "F4        0.846334         139348           8.024645e+10   \n",
       "F5        0.850245         139992           1.027483e+11   \n",
       "F6        0.817193         134550           9.062305e+10   \n",
       "F10       0.648756         106817           4.076389e+10   \n",
       "F11       0.718109         118236           5.850732e+10   \n",
       "F12       0.568689          93634           2.937665e+10   \n",
       "\n",
       "     unique_pep2_protein_count  \n",
       "F4                        9462  \n",
       "F5                        9421  \n",
       "F6                        9445  \n",
       "F10                       9239  \n",
       "F11                       9391  \n",
       "F12                       9022  \n",
       "‚ö†Ô∏è  [TrackedDataFrame] This summary has been modified and is not synced back to .obs."
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(values={'condition': ['kd','sc'], 'cellline': 'AS'})\n",
    "pdata_filtered.summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d04e8ad",
   "metadata": {},
   "source": [
    "When `exact_cases=False`, the logic is **(OR within fields, AND across fields)**.  \n",
    "This means any sample that matches *any* treatment in `['kd','sc']` **and** has `cellline='A'` is kept.\n",
    "\n",
    "| Sample | Treatment | Cell line | Match logic | Result |\n",
    "|:------:|:----------:|:----------:|:-------------|:-------:|\n",
    "| 1 | sc | A | ‚úÖ treatment in [kd, sc] and ‚úÖ cellline=A | ‚úÖ Kept |\n",
    "| 2 | kd | A | ‚úÖ treatment in [kd, sc] and ‚úÖ cellline=A | ‚úÖ Kept |\n",
    "| 3 | sc | B | ‚úÖ treatment in [kd, sc] and ‚ùå cellline=A | ‚ùå |\n",
    "| 4 | kd | B | ‚úÖ treatment in [kd, sc] and ‚ùå cellline=A | ‚ùå |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "add411a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering samples [exact match]:\n",
      "    Returning a copy of sample data based on exact match:\n",
      "     üî∏ Matching any of the following cases:\n",
      "       1. {'condition': 'kd', 'cellline': 'AS'}\n",
      "       2. {'condition': 'sc', 'cellline': 'BE'}\n",
      "     ‚ÑπÔ∏è Auto-cleanup: Removed 135 empty proteins (all-NaN or all-zero). Proteins: P04259, Q6UWP8, P23468, P48594, Q8WVV4, P22735, Q5T749, P01857, Q9C075, Q7Z3Y8, ...\n",
      "    ‚Üí Samples kept: 6, Samples dropped: 6\n",
      "    ‚Üí Proteins kept: 10258\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "      <th>duration</th>\n",
       "      <th>protein_quant</th>\n",
       "      <th>protein_count</th>\n",
       "      <th>protein_abundance_sum</th>\n",
       "      <th>mbr_count</th>\n",
       "      <th>high_count</th>\n",
       "      <th>peptide_quant</th>\n",
       "      <th>peptide_count</th>\n",
       "      <th>peptide_abundance_sum</th>\n",
       "      <th>unique_pep2_protein_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F4</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.941704</td>\n",
       "      <td>9660</td>\n",
       "      <td>8.030133e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9650</td>\n",
       "      <td>0.842202</td>\n",
       "      <td>139349</td>\n",
       "      <td>8.024650e+10</td>\n",
       "      <td>9464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F5</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.937415</td>\n",
       "      <td>9616</td>\n",
       "      <td>1.027993e+11</td>\n",
       "      <td>0</td>\n",
       "      <td>9617</td>\n",
       "      <td>0.846094</td>\n",
       "      <td>139993</td>\n",
       "      <td>1.027483e+11</td>\n",
       "      <td>9422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F6</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.940047</td>\n",
       "      <td>9643</td>\n",
       "      <td>9.066725e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9636</td>\n",
       "      <td>0.813203</td>\n",
       "      <td>134551</td>\n",
       "      <td>9.062305e+10</td>\n",
       "      <td>9447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F22</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.955352</td>\n",
       "      <td>9800</td>\n",
       "      <td>7.545933e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9803</td>\n",
       "      <td>0.752517</td>\n",
       "      <td>124510</td>\n",
       "      <td>7.548227e+10</td>\n",
       "      <td>9623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F23</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.168064</td>\n",
       "      <td>1724</td>\n",
       "      <td>1.037960e+08</td>\n",
       "      <td>0</td>\n",
       "      <td>1666</td>\n",
       "      <td>0.027348</td>\n",
       "      <td>4525</td>\n",
       "      <td>1.038191e+08</td>\n",
       "      <td>1712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F24</th>\n",
       "      <td>Sample</td>\n",
       "      <td>BE</td>\n",
       "      <td>RA</td>\n",
       "      <td>sc</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.950283</td>\n",
       "      <td>9748</td>\n",
       "      <td>7.946756e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9754</td>\n",
       "      <td>0.805304</td>\n",
       "      <td>133244</td>\n",
       "      <td>7.945111e+10</td>\n",
       "      <td>9553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample cellline treatment condition duration  protein_quant  \\\n",
       "F4   Sample       AS        RA        kd       d7       0.941704   \n",
       "F5   Sample       AS        RA        kd       d7       0.937415   \n",
       "F6   Sample       AS        RA        kd       d7       0.940047   \n",
       "F22  Sample       BE        RA        sc       d7       0.955352   \n",
       "F23  Sample       BE        RA        sc       d7       0.168064   \n",
       "F24  Sample       BE        RA        sc       d7       0.950283   \n",
       "\n",
       "     protein_count  protein_abundance_sum  mbr_count  high_count  \\\n",
       "F4            9660           8.030133e+10          0        9650   \n",
       "F5            9616           1.027993e+11          0        9617   \n",
       "F6            9643           9.066725e+10          0        9636   \n",
       "F22           9800           7.545933e+10          0        9803   \n",
       "F23           1724           1.037960e+08          0        1666   \n",
       "F24           9748           7.946756e+10          0        9754   \n",
       "\n",
       "     peptide_quant  peptide_count  peptide_abundance_sum  \\\n",
       "F4        0.842202         139349           8.024650e+10   \n",
       "F5        0.846094         139993           1.027483e+11   \n",
       "F6        0.813203         134551           9.062305e+10   \n",
       "F22       0.752517         124510           7.548227e+10   \n",
       "F23       0.027348           4525           1.038191e+08   \n",
       "F24       0.805304         133244           7.945111e+10   \n",
       "\n",
       "     unique_pep2_protein_count  \n",
       "F4                        9464  \n",
       "F5                        9422  \n",
       "F6                        9447  \n",
       "F22                       9623  \n",
       "F23                       1712  \n",
       "F24                       9553  \n",
       "‚ö†Ô∏è  [TrackedDataFrame] This summary has been modified and is not synced back to .obs."
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(\n",
    "    values=[\n",
    "        {'condition': 'kd', 'cellline': 'AS'},\n",
    "        {'condition': 'sc', 'cellline': 'BE'}\n",
    "    ],\n",
    "    exact_cases=True\n",
    ")\n",
    "pdata_filtered.summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc8298c1",
   "metadata": {},
   "source": [
    "When `exact_cases=True`, the logic requires an **exact match** to *one of the full dictionaries*.  \n",
    "Here, only samples matching either `{treatment: 'kd', cellline: 'A'}` **or** `{treatment: 'sc', cellline: 'B'}` are kept.\n",
    "\n",
    "| Sample | Treatment | Cell line | Match dictionary | Result |\n",
    "|:------:|:----------:|:----------:|:----------------:|:-------:|\n",
    "| 1 | sc | A | ‚ùå No exact match | ‚ùå |\n",
    "| 2 | kd | A | ‚úÖ Matches {'kd','A'} | ‚úÖ Kept |\n",
    "| 3 | sc | B | ‚úÖ Matches {'sc','B'} | ‚úÖ Kept |\n",
    "| 4 | kd | B | ‚ùå No exact match | ‚ùå |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d117b4c5",
   "metadata": {},
   "source": [
    "### Filter by condition\n",
    "\n",
    "Use a logical condition string referencing columns in `pdata.summary`.  \n",
    "This enables numeric and boolean filtering based on sample-level summary statistics.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "31584112",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering samples [condition]:\n",
      "    Returning a copy of sample data based on condition:\n",
      "     üî∏ Condition: protein_count > 1000\n",
      "     ‚ÑπÔ∏è Auto-cleanup: No empty proteins found (all-NaN or all-zero).\n",
      "    ‚Üí Samples kept: 12, Samples dropped: 0\n",
      "    ‚Üí Proteins kept: 10393\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(condition=\"protein_count > 1000\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b9a2238",
   "metadata": {},
   "source": [
    "#### Using `min_prot`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e2bb04ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering samples [condition]:\n",
      "    Returning a copy of sample data based on condition:\n",
      "     üî∏ Condition: protein_count >= 1000\n",
      "     ‚ÑπÔ∏è Auto-cleanup: No empty proteins found (all-NaN or all-zero).\n",
      "    ‚Üí Samples kept: 12, Samples dropped: 0\n",
      "    ‚Üí Proteins kept: 10393\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(min_prot=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae7d4b38",
   "metadata": {},
   "source": [
    "### Filter by file list\n",
    "\n",
    "Filter samples directly by their file or sample identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "f9445394",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering samples [file list]:\n",
      "    Returning a copy of sample data based on file list:\n",
      "     üî∏ Files requested (including): 2\n",
      "     üî∏ Missing samples ignored: 1\n",
      "     ‚ÑπÔ∏è Auto-cleanup: Removed 733 empty proteins (all-NaN or all-zero). Proteins: Q12756, O15523, O00507, P51531, Q02413, Q8WWN8, Q9ULE6, P58107, P20930, P08779, ...\n",
      "    ‚Üí Samples kept: 1, Samples dropped: 11\n",
      "    ‚Üí Proteins kept: 9660\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(file_list=['F4', 'F7'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e10b9c87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß≠ [USER] Filtering samples [file list]:\n",
      "    Returning a copy of sample data based on file list:\n",
      "     üî∏ Files requested (excluding): 2\n",
      "     üî∏ Missing samples ignored: 1\n",
      "     ‚ÑπÔ∏è Auto-cleanup: Removed 1 empty proteins (all-NaN or all-zero). Proteins: P0C0L5\n",
      "    ‚Üí Samples kept: 11, Samples dropped: 1\n",
      "    ‚Üí Proteins kept: 10392\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(exclude_file_list=['F4', 'F7'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9997332",
   "metadata": {},
   "source": [
    "### Advanced query mode\n",
    "\n",
    "Enable **advanced filtering** with `query_mode=True` to execute raw pandas-style queries.  \n",
    "This interprets `values` or `condition` as a raw `.query()` string evaluated directly on `.obs` or `.summary`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "bbd65c2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "‚ö†Ô∏è [WARN] Advanced query mode enabled ‚Äî interpreting string as a pandas-style expression.\n",
      "üß≠ [USER] Filtering samples [query]:\n",
      "    Returning a copy of sample data based on query string:\n",
      "   üî∏ Query: cellline == 'AS' and condition == 'kd'\n",
      "     ‚ÑπÔ∏è Auto-cleanup: Removed 492 empty proteins (all-NaN or all-zero). Proteins: O15523, O00507, P51531, Q02413, P20930, P08779, Q8TDI0, P04259, A8K2U0, O76038, ...\n",
      "    ‚Üí Samples kept: 3, Samples dropped: 9\n",
      "    ‚Üí Proteins kept: 9901\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "      <th>duration</th>\n",
       "      <th>protein_quant</th>\n",
       "      <th>protein_count</th>\n",
       "      <th>protein_abundance_sum</th>\n",
       "      <th>mbr_count</th>\n",
       "      <th>high_count</th>\n",
       "      <th>peptide_quant</th>\n",
       "      <th>peptide_count</th>\n",
       "      <th>peptide_abundance_sum</th>\n",
       "      <th>unique_pep2_protein_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F4</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.975659</td>\n",
       "      <td>9660</td>\n",
       "      <td>8.030133e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9639</td>\n",
       "      <td>0.852406</td>\n",
       "      <td>139347</td>\n",
       "      <td>8.024644e+10</td>\n",
       "      <td>9465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F5</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.971215</td>\n",
       "      <td>9616</td>\n",
       "      <td>1.027993e+11</td>\n",
       "      <td>0</td>\n",
       "      <td>9605</td>\n",
       "      <td>0.856345</td>\n",
       "      <td>139991</td>\n",
       "      <td>1.027482e+11</td>\n",
       "      <td>9422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F6</th>\n",
       "      <td>Sample</td>\n",
       "      <td>AS</td>\n",
       "      <td>RA</td>\n",
       "      <td>kd</td>\n",
       "      <td>d7</td>\n",
       "      <td>0.973942</td>\n",
       "      <td>9643</td>\n",
       "      <td>9.066725e+10</td>\n",
       "      <td>0</td>\n",
       "      <td>9627</td>\n",
       "      <td>0.823062</td>\n",
       "      <td>134550</td>\n",
       "      <td>9.062305e+10</td>\n",
       "      <td>9448</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample cellline treatment condition duration  protein_quant  \\\n",
       "F4  Sample       AS        RA        kd       d7       0.975659   \n",
       "F5  Sample       AS        RA        kd       d7       0.971215   \n",
       "F6  Sample       AS        RA        kd       d7       0.973942   \n",
       "\n",
       "    protein_count  protein_abundance_sum  mbr_count  high_count  \\\n",
       "F4           9660           8.030133e+10          0        9639   \n",
       "F5           9616           1.027993e+11          0        9605   \n",
       "F6           9643           9.066725e+10          0        9627   \n",
       "\n",
       "    peptide_quant  peptide_count  peptide_abundance_sum  \\\n",
       "F4       0.852406         139347           8.024644e+10   \n",
       "F5       0.856345         139991           1.027482e+11   \n",
       "F6       0.823062         134550           9.062305e+10   \n",
       "\n",
       "    unique_pep2_protein_count  \n",
       "F4                       9465  \n",
       "F5                       9422  \n",
       "F6                       9448  \n",
       "‚ö†Ô∏è  [TrackedDataFrame] This summary has been modified and is not synced back to .obs."
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(values=\"cellline == 'AS' and condition == 'kd'\", query_mode=True)\n",
    "pdata_filtered.summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c8e66e57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "‚ö†Ô∏è [WARN] Advanced query mode enabled ‚Äî interpreting string as a pandas-style expression.\n",
      "üß≠ [USER] Filtering samples [query]:\n",
      "    Returning a copy of sample data based on query string:\n",
      "   üî∏ Query: protein_count > 1000 and protein_quant > 0.9\n",
      "     ‚ÑπÔ∏è Auto-cleanup: Removed 2 empty proteins (all-NaN or all-zero). Proteins: P23468, Q06187\n",
      "    ‚Üí Samples kept: 10, Samples dropped: 2\n",
      "    ‚Üí Proteins kept: 10391\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_sample(condition=\"protein_count > 1000 and protein_quant > 0.9\", query_mode=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ef377f",
   "metadata": {},
   "source": [
    "## `filter_rs()`\n",
    "\n",
    "Filter the RS matrix and associated `.prot` and `.pep` data based on peptide‚Äìprotein relationships.  \n",
    "This method applies rules for retaining proteins with sufficient peptide evidence and/or removing ambiguous peptides.\n",
    "\n",
    "### Key Parameters\n",
    "\n",
    "- **`min_peptides_per_protein`** *(int, optional)* ‚Äì Minimum total number of peptides required per protein.  \n",
    "- **`min_unique_peptides_per_protein`** *(int, optional)* ‚Äì Minimum number of *unique* peptides required per protein.  \n",
    "- **`max_proteins_per_peptide`** *(int, optional)* ‚Äì Maximum number of proteins a peptide can map to (peptides exceeding this are removed).  \n",
    "- **`preset`** *(str or dict, optional)* ‚Äì Predefined filter presets:  \n",
    "  - `\"default\"` ‚Üí unique peptides ‚â• 2  \n",
    "  - `\"lenient\"` ‚Üí total peptides ‚â• 2  \n",
    "  - A dictionary specifying thresholds manually.  \n",
    "  The default preset is `\"default\"`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1b4a9f5",
   "metadata": {},
   "source": [
    "### Filter by unique peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c40ebf98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß™ Filtered RS, min unique peptides: 2. Proteins: 10393 ‚Üí 10068 (dropped 325), Peptides: 167114 ‚Üí 165468 (dropped 1646).\n",
      "‚ÑπÔ∏è RS matrix: (10068, 165468) (proteins √ó peptides), sparsity: 99.99%\n",
      "   - Proteins with ‚â•2 *unique* linked peptides: 10068/10068\n",
      "   - Peptides linked to ‚â•2 proteins: 4452/165468\n",
      "   - Mean peptides per protein: 16.98\n",
      "   - Mean proteins per peptide: 1.03\n",
      "     ‚úÖ [OK] pAnnData object is valid.\n"
     ]
    }
   ],
   "source": [
    "pdata_filtered = pdata.filter_rs(min_unique_peptides_per_protein=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7b79d2b",
   "metadata": {},
   "source": [
    " Protein | Peptide 1 | Peptide 2 | Peptide 3 | Unique peptides | Result |\n",
    "|:--------|:----------:|:----------:|:----------:|:----------------:|:-------:|\n",
    "| P001 | üü© | üü© | ‚ö™ | 2 | ‚úÖ Kept |\n",
    "| P002 | üü© | ‚ö™ | ‚ö™ | 1 | ‚ùå Filtered |\n",
    "| P003 | üü© | üü© | üü© | 3 | ‚úÖ Kept |\n",
    "\n",
    "> Proteins with **fewer than two unique peptides** are removed by default.  \n",
    "> The filtering operation updates both `.prot` and `.pep` tables and synchronizes their mappings in the RS matrix."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
